### PR TITLE
implement supervisor status command

### PR DESCRIPTION
**Pull request by mwrock**
This adds a `status command to the supervisor. It takes an optional package ident argument. If this argument is not given then the command reports status for all services running in the supervisor.

Further yhis fixes a bug in the service data persistence where service state was not being persisted when a service's state changes. In production, the `/services` monitoring endpoint only reports the initial service state when the supervisor first runs and this would typically report that the services are `Down` with a `Null` PID. Now service states are checked for a change on each supervisor tick and persisted to disk on change.

Note that the `state_entered` supervisor field is now a `Timespec` rather than a `SteadyTime`. This is because `SteadyTime` does not lend itself well  to serialization/deserialization. Its internal state will differ based on platform whereas `Timespec` internal data should be uniform.

Signed-off-by: Matt Wrock <matt@mattwrock.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by mwrock** [comment]
Status for an individual service:
```
C:\dev\habitat\components\sup [status +1 ~2 -0 !]> ..\..\target\debug\hab-sup.exe status core/habitat-aspnet-sample
hab-sup(MN): The core/habitat-aspnet-sample/0.2.0/20170311090829 service entered the running state with PID: 15492 11 seconds ago.
```

Status for all services:
```
C:\dev\habitat\components\sup [status +1 ~2 -0 !]> ..\..\target\debug\hab-sup.exe status

hab-sup(MN): The core/habitat-aspnet-sample/0.2.0/20170311090829 service entered the running state with PID: 15492 83 seconds ago.
hab-sup(MN): The core/mysql/5.7.17/20170310233054 service entered the running state with PID: 4660 5 seconds ago.
```

**Comment by mwrock** [comment]
Thanks a ton for this feedback @smith!
 Here is what I propose for exit codes:

`0` - single service running or `status` with no ident is passed and the supervisor is running.
`1` - single service `down`
`2` - single service `starting`
`3` - single service restarting
`4` - single service not loaded
`5` - supervisor not running

Thoughts?

**Comment by mwrock** [comment]
Having a hard time getting a repro on your second point @smith:
```
C:\dev\habitat\components\sup [status +1 ~0 -0 !]> ..\..\target\debug\hab-sup.exe status core/mysql
hab-sup(MN): The core/mysql service is not currently loaded.
C:\dev\habitat\components\sup [status +1 ~0 -0 !]> ..\..\target\debug\hab-sup.exe load core/mysql
C:\dev\habitat\components\sup [status +1 ~0 -0 !]> ..\..\target\debug\hab-sup.exe status core/mysql
hab-sup(MN): The core/mysql/5.7.17/20170310233054 service is not currently running.
C:\dev\habitat\components\sup [status +1 ~0 -0 !]> ..\..\target\debug\hab-sup.exe status core/mysql
hab-sup(MN): The core/mysql/5.7.17/20170310233054 service entered the running state with PID: 5024 2 seconds ago.
```

One important not for testing is that you need to start the supervisor (via run or start) with the new binary. If you the supervisor is running from the prod binary, it will not refresh the statuses and the `status` command of the build binary may be stale.

**Comment by mwrock** [comment]
ok I think I have a repro now. if i load it, stop it, kill the supervisor and then start the supervisor with a different service. The mysql service is stil loaded but stopped yet dead to the status command.

**Comment by mwrock** [comment]
@smith I *think* I hit all of your feedback but would really appretiate it if you could run the latest in this PR and make sure I addressed the second point where it reported that the loaded service was not loaded.

**Comment by mwrock** [comment]
Note that we used to only add services that had an initial `desired_state` of `Up` to the manager's service collection. This meant that out `/services` api did not report services that were loaded and stopped. We now load all services regardless of desired state. If the `desired_state` is `Down` we simply break out of that service's `tick` function early so it does nothing.

File: components/sup/src/manager/mod.rs [hunk 1]
```diff
@@ -156,6 +206,26 @@ impl Manager {
...
+
+        Err(sup_error!(Error::ServiceNotLoaded(ident)))
+    }
+
+    pub fn status(cfg: ManagerConfig) -> Result<Vec<ServiceStatus>> {
+        let state_path = Self::state_path_from(&cfg);
+        let fs_cfg = FsCfg::new(state_path);
+
+        let dat = File::open(&fs_cfg.services_data_path)?;
+        Ok(serde_json::from_reader(&dat)?)
```
Comment (reset) [inline components/sup/src/manager/mod.rs]: Directly converting one error to our own type is a pattern from the earlier days of our codebase but it's something that we've been trying to get away from. It's more appropriate to create a specific error and pass the error you would be converting here instead directly to it as a member. For example, you'd maybe create the error `SvcStatFileParse(serde_json::Error)` which explicitly says that there was an error in parsing the svc stat file. Since you're holding onto the `serde_json::Error` in the error struct you can display that in anyway you'd like, including just writing it into your error string at the end

File: components/sup/src/manager/service/mod.rs [hunk 2]
```diff
@@ -287,6 +289,9 @@ impl Service {
     }
 
     pub fn tick(&mut self, census_ring: &CensusRing) -> bool {
+        if self.desired_state == DesiredState::Down {
```
Comment (reset) [inline components/sup/src/manager/service/mod.rs]: This shouldn't be a necessary. Services which are in the `Down` state shouldn't be loaded so their tick function shouldn't be called

File: components/sup/src/main.rs [hunk 3]
```diff
@@ -314,6 +324,73 @@ fn sub_start(m: &ArgMatches) -> Result<()> {
...
+                    outputln!("The {} service entered the restarting state {} seconds ago.",
+                              status.package,
+                              status.supervisor.elapsed.num_seconds());
+                    if query_one {
+                        std::process::exit(4);
+                    }
+                }
+            }
+        }
+    } else {
```
Comment (reset) [inline components/sup/src/main.rs]: Instead of an `if`/`else` block I'd recommend using a _function short-circuit_ where you return out of the function early.

```rust
if !Manager::is_running(&cfg)? {
    std::process::exit(6);
}

// ... rest of code
```

This would be a design pattern that we have in our coding standards had anyone ever gotten around to writing them yet ;)

File: components/sup/src/manager/service/mod.rs [hunk 4]
```diff
@@ -92,7 +93,7 @@ pub struct Service {
     last_health_check: Instant,
     #[serde(skip_serializing)]
     manager_fs_cfg: Arc<manager::FsCfg>,
-    supervisor: Supervisor,
+    pub supervisor: Supervisor,
```
Comment (reset) [inline components/sup/src/manager/service/mod.rs]: We should keep this private and expose some public member functions on the `Service` struct to expose the API that you need. The `Supervisor` module should be considered internal

File: components/sup/src/supervisor.rs [hunk 5]
```diff
@@ -104,7 +104,7 @@ pub struct Supervisor {
     pub package: Arc<RwLock<PackageInstall>>,
     pub preamble: String,
     pub state: ProcessState,
-    pub state_entered: SteadyTime,
+    pub state_entered: Timespec,
```
Comment (reset) [inline components/sup/src/supervisor.rs]: From my understanding you had to make this swap because of serialization issues. It's been a while since we've had to revisit this module so this hasn't come up before, but the crate that we're using for time is actually old and deprecated which is why you don't have the support you need in it for serde.

[chrono](https://github.com/chronotope/chrono) replaced the [time](https://github.com/rust-lang-deprecated/time) crate and it does have support for serialization using serde. You might want to consider using that here instead or we could save that for a future refactor!

File: components/sup/src/manager/mod.rs [hunk 6]
```diff
@@ -424,6 +495,10 @@ impl Manager {
                                     &self.butterfly.service_file_store);
             service_rumor_offset = 0;
 
+            if self.check_for_changed_services() {
```
Comment (reset) [inline components/sup/src/manager/mod.rs]: Good catch!

File: components/sup/src/manager/mod.rs [hunk 7]
```diff
@@ -650,12 +741,7 @@ impl Manager {
     fn start_initial_services_from_watcher(&mut self) -> Result<()> {
         for service_event in self.watcher.initial_events()? {
             match service_event {
-                SpecWatcherEvent::AddService(spec) => {
```
Comment (reset) [inline components/sup/src/manager/mod.rs]: I don't think we should load services which are supposed to be in the `Down` state, I'd keep this block as is. The code makes assumptions that a "loaded" service has an active rumor which is shared along with participating in the tick loop. I think this complicates that logic by needing to short circuit things with a state check.

It's worth noting that right now Adam is working on some logic which will allow us to tombstone service rumors so when you stop a service it actually let's peers know that it went away

File: components/sup/src/main.rs [hunk 8]
```diff
@@ -187,6 +189,14 @@ fn cli<'a, 'b>() -> App<'a, 'b> {
             (@arg CONFIG_DIR: --("config-from") +takes_value {dir_exists}
                 "Use package config from this path, rather than the package itself")
         )
+        (@subcommand status =>
+            (about: "Query the status of a running Habitat service.")
```
Comment (reset) [inline components/sup/src/main.rs]: Technically you're not just asking about _running_ services ;)

File: components/sup/src/manager/mod.rs (hunk 7 above)
Comment (reset) [inline components/sup/src/manager/mod.rs]: You may want to read the service specs directory to get a view of what is stopped vs what is running. As @smith pointed out, things can be "loaded" but not be running

Comment (mwrock) [inline components/sup/src/manager/mod.rs]: This was mainly a choice of where to put the complexity. If the `services` vec only has services with a `DesiredState` of `Up` then we need to change the persistence logic to do more than just iterate the `services` vec. If the "downed"services are not in the services vec, they won't get hydrated to the serialized services. We could aggregate and merge the down and up services into a new data structure that we serialize to `/services` but that seems even more complex.

File: components/sup/src/manager/mod.rs (hunk 1 above)
Comment (mwrock) [inline components/sup/src/manager/mod.rs]: Is this what
```
impl From<serde_json::Error> for SupError {
 +    fn from(err: serde_json::Error) -> SupError {
 +        sup_error!(Error::SerdeJsonError(err))
 +    }
 +}
```
is doing?

Comment (mwrock) [inline components/sup/src/manager/mod.rs]: Oh I think i see...

You are saying bail on that `From<serde_json::Error>` implementation and actually match here and explicitly `Err` my `sup_error!(Error::SerdeJsonError(err))`

That does sound better

File: components/sup/src/supervisor.rs (hunk 5 above)
Comment (mwrock) [inline components/sup/src/supervisor.rs]: Yeah I stumbled upon chrono in the serde issues and decided to skip the dependency since I would have to do some custom deserialization anyway to represent the time of the state change as elapsed seconds.

**Comment by smith** [comment]
This is much closer, but a couple things:

If I stop a transient service, it seems like it still stays in the list. Can you reproduce this? If there's no spec file it shouldn't show up:

```bash
[27][default:/src:130]# hab pkg exec smith/hab-sup hab-sup status
hab-sup(MN): The core/postgresql/9.6.1/20170424032128 service is not currently running.
hab-sup(MN): The core/redis/3.2.4/20170215222111 service is not currently running.
[28][default:/src:0]# ls /hab/sup/default/specs/
postgresql.spec
```

Also, I'm wondering about the correct exit code behavior. The first thing I want to do with this is:

* Is service x loaded? (I don't care whether or not it's running)
  * If not, `hab sup load x`
  * If so, `hab sup start x`

basically, I want to do `hab sup load --force` but not force a restart.

In this case I would want `hab sup status x`, when x is persistent but down, to exit 0, but in the case where I just wanted to know if it was up or down, I would probably expect a non-zero exit code.

This confusion makes me think either there's another missing feature (`hab sup load --yes-I-want-it-loaded-but-if-it's-already-loaded-then-thats-cool-never-mind`?) or some other simplification that could be done. Thoughts?

**Comment by mwrock** [comment]
@smith I'm just pushed a new iteration with some refactorings based on comments from @reset. They had the fortunate byproduct of also fixing the issue with unloaded services showing up as not running in the status.

Thinking of your exit code comments, I do think that persistent services not running should emit a `0` exit code since that is not an "error" case. If it mattered to you that the service was running you could grep the output. I think simply changing the error of loading a loaded service to a warning would be best. We can open a separate issue for that.

**Comment by mwrock** [comment]
Hmmm. Now I'm feeling torn over this exit code business. If a stopped transient service returns 0 because that is an acceptable state, then we should also return 0 got services in the "starting" and "reloading" states. You could of coarse grep on any of these states reliably, but status codes are nice. Here is what I think we should do: remove the exit codes for "acceptable" states and if that proves inconvenient, we later add a `-states` arg that will provide explicit exit codes for each possible state.

**Comment by reset** [comment]
@mwrock when you stop a transient service it unloads from the supervisor completely - the service is stopped and the service spec should be deleted. I don't think you can ever see a transiet service as "stopped"

**Comment by mwrock** [comment]
@reset yeah @smith was saying that he was seeing that in the previous state of this PR but I have not been able to repro and it would certainly be a bug

**Comment by reset** [comment]
tenor-67443071

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/226929712).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/226929712).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265
